### PR TITLE
fix(perf): Fix performance bug where maxPages=0 caused lock up

### PR DESCRIPTION
**Pull request by ragaeeb**
# Performance Test Suite for Segmentation Library

### [DELETE] [perf-test.ts](perf-test.ts)

---

### [NEW] `src/segmentation/segmenter.perf.test.ts`

Uses `RUN_PERF=true` env var to skip by default (won't run with regular `bun test`):

```typescript
const perfDescribe = process.env.RUN_PERF === 'true' ? describe : describe.skip;

perfDescribe('Performance Tests', () => {
  const pages = generateArabicPages(10_000);
  const CI_MULTIPLIER = process.env.CI ? 2 : 1;
  
  // ... tests with thresholds multiplied by CI_MULTIPLIER
});
```

#### Test Matrix (25+ tests)

**maxPages**: 0, 1, 2, 20  
**Rule types**: lineStartsWith (single & large array), lineEndsWith, lineStartsAfter, regex, template, fuzzy  
**Large rule sets**: 50 regex, 50 template, 50 mixed  
**maxContentLength**: 500, 2000  
**Breakpoints**: many patterns, prefer=shorter  
**Exclusions**: 5000 pages  
**Split behavior**: occurrence=all/first/last

---

### [MODIFY] [package.json](package.json)
```diff
 "scripts": {
     "build": "tsdown",
-    "perf": "bun run perf-test.ts"
+    "test": "bun test",
+    "test:perf": "RUN_PERF=true bun test src/segmentation/segmenter.perf.test.ts"
 }
```

---

### [MODIFY] [.github/workflows/build.yml](.github/workflows/build.yml)
```yaml
- name: Performance Tests
  run: RUN_PERF=true CI=true bun test src/segmentation/segmenter.perf.test.ts
```

# Performance Test Suite - Walkthrough

## Changes Made

| File | Action |
|------|--------|
| [segmenter.perf.test.ts](src/segmentation/segmenter.perf.test.ts) | NEW - 22 performance tests |
| [perf-test.ts](perf-test.ts) | DELETED |
| [package.json](package.json) | Added [test](src/index.test.ts#7-21) and `test:perf` scripts |
| [build.yml](.github/workflows/build.yml) | Added performance test step |

## Test Results (22 tests, 2.55s total)

```
✓ maxPages=0 [8.89ms]
✓ maxPages=1 [11.11ms]
✓ maxPages=2 [11.86ms]
✓ maxPages=20 [13.34ms]
✓ lineStartsWith (single) [483ms]
✓ lineStartsWith (20 patterns) [78ms]
✓ lineEndsWith [80ms]
✓ lineStartsAfter [57ms]
✓ regex rule [27ms]
✓ template rule [169ms]
✓ fuzzy lineStartsWith [479ms]
✓ 50 regex rules [26ms]
✓ 50 template rules [24ms]
✓ 50 mixed rules [319ms]
✓ maxContentLength 500 [105ms]
✓ maxContentLength 2000 [46ms]
✓ many breakpoints [8ms]
✓ prefer=shorter [10ms]
✓ large exclude set [10ms]
✓ occurrence=all [477ms]
✓ occurrence=first [34ms]
✓ occurrence=last [33ms]
```

## Usage

```bash
# Run performance tests
bun run test:perf

# Regular tests skip perf tests
bun test
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Reduced page-boundary false positives and tightened fallbacks to avoid cross-page merges.

* **Performance**
  * Fast-path for very large documents (1000+ pages) to speed boundary calculations and oversized-segment handling; improved perf logging and page-join normalization.

* **Tests**
  * Added a gated performance test suite, removed the old perf script, and added a CI performance step.

* **Documentation**
  * Added ASCII-digit tokens (num/nums) and guidance for choosing ASCII vs Arabic-Indic numerals.

* **Chores**
  * Bumped package version and updated test scripts.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by ragaeeb** [comment]
### Amazon Nova 2 Pro AI Code Review: Segmentation Performance Fix

---

## ✅ **Overall Assessment**

**Great job!** This is a **robust fix** for a severe performance regression. The core insight — that O(n²) string-search in `buildBoundaryPositions()` was the bottleneck for large books — is correctly identified and solved with a **fast path for 1000+ page segments**.

The implementation is:
- **Correct**: Handles all edge cases (non-contiguous IDs, missing pages)
- **Safe**: Falls back to accurate string-search for small segments
- **Well-tested**: Comprehensive performance test suite prevents regressions
- **Maintainable**: Clear separation of fast vs accurate paths

---

## 1. **Algorithm Correctness**

### ✅ **1000-page threshold is appropriate**
- **Why it works**: 
  - For **<1000 pages**: Accurate string-search handles offset drift from structural rules (critical for correctness)
  - For **≥1000 pages**: Cumulative offsets are stable enough; performance gain outweighs minor inaccuracy
- **Data point**: 25k pages went from **hang → 160ms** — threshold is in the right order of magnitude

### ✅ **`effectiveMaxPages = maxPages + 1` is correct**
| `maxPages` | Meaning | `effectiveMaxPages` | Example |
|------------|---------|---------------------|---------|
| `0` | 1 page per segment | `1` | Pages 0-0 |
| `1` | 2 pages per segment | `2` | Pages 0-1 |
| `20` | 21 pages per segment | `21` | Pages 0-20 |

This correctly implements **"up to and including"** page counting.

### ⚠️ **Edge Case: Non-contiguous page IDs**
Your fix handles this perfectly:
```typescript
// In fast path:
const boundary = Math.max(0, (cumulativeOffsets[i] ?? 0) - baseOffset);
const prevBoundary = boundaryPositions[boundaryPositions.length - 1];
boundaryPositions.push(Math.max(prevBoundary + 1, Math.min(boundary, segmentContent.length)));
```

- Uses **actual page IDs** (`pageIds[]`) for slicing
- No assumption of sequential IDs (`0,1,2...`)
- ✅ **Passed in tests** with gaps (pages 59 → 229)

---

## 2. **Missed Performance Issues**

### ✅ **No major misses**
All critical paths are covered:
1. `buildBoundaryPositions()` → **Fixed**
2. `processOversizedSegment()` → **Fixed**
3. `applyPageJoinerBetweenPages()` → **Not a bottleneck** (only runs on output segments, not during processing)

### 🔍 **Minor optimization opportunity**
In `findPageStartNearExpectedBoundary()`, the **dual search** (forward + backward) could be optimized for large segments by:
```typescript
// Instead of searching full 20k window, search only near expected boundary
const SEARCH_WINDOW = 5000; // Instead of 20k
const searchStart = Math.max(0, approx - SEARCH_WINDOW);
const searchEnd = Math.min(remainingContent.length, approx + SEARCH_WINDOW);
```
But this is **optional** — current fix already solves the main regression.

---

## 3. **Test Coverage Gaps**

### ✅ **Excellent coverage**
Your performance test suite is **comprehensive**:
| Configuration | Covered? |
|---------------|----------|
| `maxPages=0` | ✅ |
| `maxPages=1` | ✅ |
| `maxPages=20` | ✅ |
| Single rule types | ✅ |
| Large rule sets (50 rules) | ✅ |
| `maxContentLength` | ✅ |
| Breakpoints + `prefer` | ✅ |
| Exclusions | ✅ |
| Occurrence variants | ✅ |

### 💡 **Suggested additions**
Add these to `segmenter.perf.test.ts`:

#### 3.1. **Edge-case pages**
```typescript
it('should handle empty pages', () => {
    const pages = generateArabicPages(5000);
    // Set 50% of pages to empty content
    pages.forEach((p, i) => { if (i % 2 === 0) p.content = ''; });
    const result = segmentPages(pages, { breakpoints: [''], maxPages: 0 });
    expect(result.length).toBeGreaterThan(0);
});
```

#### 3.2. **Single-character pages**
```typescript
it('should handle single-character pages', () => {
    const pages = Array.from({ length: 10000 }, (_, i) => ({
        content: 'ا',
        id: i + 1
    }));
    const result = segmentPages(pages, { breakpoints: [''], maxPages: 0 });
    expect(result.length).toBe(10000);
});
```

#### 3.3. **Memory usage test**
Add a memory test to catch leaks:
```typescript
it('should not leak memory with 25k pages', async () => {
    const pages = generateArabicPages(25000);
    const memoryBefore = process.memoryUsage().heapUsed;
    segmentPages(pages, { breakpoints: [''], maxPages: 0 });
    const memoryAfter = process.memoryUsage().heapUsed;
    expect(memoryAfter - memoryBefore).toBeLessThan(50 * 1024 * 1024); // <50MB
});
```

---

## 4. **Code Quality**

### ✅ **Fast-path logic is clear**
The fast-path code is **well-commented** and **easy to understand**:
```typescript
// FAST PATH: For large segments (1000+ pages), use cumulative offsets directly
if (pageCount >= FAST_PATH_THRESHOLD && !maxContentLength && !debugMetaKey) {
    logger?.debug?.('[breakpoints] Using offset-based fast-path...');
    // ... simple arithmetic ...
}
```

### ✅ **Debug logs are sufficient**
You have:
- `logger?.debug?.()` for key decisions
- `logger?.trace?.()` for per-iteration details (disabled by default)
- **No logs inside tight loops** (critical for performance)

### ✅ **`describe.skip` pattern is correct**
Using `process.env.RUN_PERF === 'true' ? describe : describe.skip` is **standard practice** for optional performance tests.

---

## 5. **Potential Regressions**

### ✅ **No regressions identified**
All edge cases are handled:
| Scenario | Handling |
|----------|----------|
| `debugMetaKey` set | Skips fast path (preserves provenance) |
| `maxContentLength` set | Skips fast path (needs character accuracy) |
| Non-contiguous page IDs | Uses `pageIds[i]` directly |
| Missing pages in `cumulativeOffsets` | Uses `?? 0` fallback |

### 🔍 **One subtle case to test**
What if `maxPages=0` **and** `pageIds` has **duplicate IDs**?
```typescript
// Test case
const pages = [
    { content: 'Page 1', id: 1 },
    { content: 'Page 1', id: 1 } // Duplicate ID
];
```
Your code uses `pageIds[i]`, which would process the same ID twice — but this is **data error**, not code bug. Add a test to document behavior.

---

## 🏆 **Recommendations**

### 1. **Add the edge-case tests** (empty pages, single-char pages, memory)
These will make the test suite **bulletproof**.

### 2. **Document the 1000-page threshold**
Add to JSDoc in `breakpoint-utils.ts`:
```typescript
/**
 * Threshold for using offset-based fast path in buildBoundaryPositions.
 * Below this: accurate string-search (handles offset drift).
 * Above this: O(n) arithmetic (performance critical).
 */
const FAST_PATH_THRESHOLD = 1000;
```

### 3. **Consider making threshold configurable**
For power users:
```typescript
interface SegmentationOptions {
    // Optional: override fast-path threshold
    fastPathThreshold?: number; // Default: 1000
}
```

---

## 📊 **Performance Results Summary**

| Configuration | Before | After |
|---------------|--------|-------|
| `maxPages=0`, 25k pages | **∞ (hang)** | **160ms** |
| `maxPages=1`, 25k pages | **∞ (hang)** | **177ms** |
| `maxPages=20`, 10k pages | **Slow** | **13ms** |

**Throughput**: ~150k pages/second — **excellent** for Arabic text segmentation.

---

## ✅ **Final Verdict**

**This fix is production-ready.** It solves the performance regression **correctly**, **safely**, and with **excellent test coverage**. Ship it! 🚀

**Only minor improvements suggested** (edge-case tests, documentation). No critical issues found.

**Comment by ragaeeb** [comment]
# Claude Sonnet 4.5

This is an excellent performance fix with thorough documentation! I'll provide detailed feedback on each of your questions and flag some potential issues.

## 🚨 Critical Issues

### 1. **Page Count vs Content Span Mismatch**

**Location**: Both fast paths use `pageCount = toIdx - fromIdx + 1`

**Problem**: This counts *index positions*, not the actual span of page IDs. With non-contiguous page IDs (which you mentioned in the book data), this creates a semantic mismatch:

```typescript
// Pages: [1, 5, 100, 1000, 1001, ..., 2000]
// If segment spans pages 1-1000:
pageCount = 1000 - 0 + 1 = 1001 ✓ (correct)

// But if segment spans pages 1000-2000:
pageCount = 1001 - 1000 + 1 = 2 ✗ (wrong! actually 1001 pages)
```

**Fix**: Use the actual page ID span instead:
```typescript
const pageSpan = pageIds[toIdx] - pageIds[fromIdx]; // page ID difference
const FAST_PATH_THRESHOLD = 1000;
if (pageSpan >= FAST_PATH_THRESHOLD && !maxContentLength && !debugMetaKey) {
```

This ensures the fast path activates based on the *semantic* page span, not the array index span.

---

### 2. **Cumulative Offset Safety**

**Location**: `breakpoint-utils.ts:446-448`

**Problem**: No validation that `cumulativeOffsets[i]` exists:

```typescript
const boundary = Math.max(0, (cumulativeOffsets[i] ?? 0) - baseOffset);
```

If `cumulativeOffsets` is shorter than expected or has undefined entries, this silently defaults to `0`, which could create duplicate boundaries at position 0.

**Fix**: Add validation before the fast path:
```typescript
if (pageCount >= FAST_PATH_THRESHOLD) {
    // Validate cumulative offsets exist for entire range
    if (cumulativeOffsets.length < toIdx + 1) {
        logger?.warn?.('[breakpoints] Cumulative offsets incomplete, falling back to accurate path');
        // Fall through to string-search path
    } else {
        // Fast path...
    }
}
```

---

### 3. **Missing Edge Case: Empty Pages**

**Location**: `breakpoint-processor.ts:278-295`

**Problem**: Fast path assumes all content slices will be non-empty:

```typescript
const content = fullContent.slice(startOffset, endOffset).trim();
if (content) {
    result.push({ content, from: pageIds[segStart], to: pageIds[segEnd] });
}
```

But what if `startOffset === endOffset` due to empty pages or stripped markers? The fast path skips these, while the slow path might handle them differently, creating inconsistent behavior.

**Test needed**:
```typescript
it('handles consecutive empty pages in fast path', () => {
    const pages: Page[] = Array.from({ length: 2000 }, (_, i) => ({
        id: i + 1,
        content: i % 100 === 0 ? '' : 'text', // Every 100th page is empty
    }));
    const result = segmentPages(pages, { maxPages: 0, breakpoints: [''], rules: [] });
    expect(result.every(s => s.content.length > 0)).toBe(true);
});
```

---

## Answers to Your Questions

### 1. **Algorithm Correctness**

**Q: Is the 1000-page threshold appropriate?**

**A**: Yes for typical books, but I'd make it configurable:

```typescript
const FAST_PATH_THRESHOLD = options.fastPathThreshold ?? 1000;
```

Add to `SegmentationOptions`:
```typescript
/**
 * Minimum page span to activate fast-path optimizations.
 * Lower values favor performance; higher values favor accuracy.
 * @default 1000
 */
fastPathThreshold?: number;
```

**Q: Edge cases where fast path produces incorrect results?**

**A**: Yes, several:

1. **Stripped markers**: If structural rules use `lineStartsAfter` and strip content, the fast path's offset-based calculation will be wrong because it doesn't account for the removed bytes. 
   
   **Solution**: Document this clearly or add a runtime check:
   ```typescript
   const hasLineStartsAfter = rules.some(r => 'lineStartsAfter' in r);
   if (hasLineStartsAfter && pageCount >= FAST_PATH_THRESHOLD) {
       logger?.warn?.('[breakpoints] Fast path may be inaccurate with lineStartsAfter rules');
   }
   ```

2. **HTML stripping**: If using `stripHtmlTags()` or the `replace` option before segmentation, offsets will drift.

**Q: Is `effectiveMaxPages = maxPages + 1` correct?**

**A**: Yes! Your comment is helpful. Consider renaming:
```typescript
const pagesPerSegment = maxPages + 1; // maxPages=0 → 1 page, maxPages=1 → 2 pages, etc.
```

---

### 2. **Missed Performance Issues**

**Q: Other O(n²) or O(n×m) patterns?**

**A**: Yes, found one:

**In `applyPageJoinerBetweenPages` (breakpoint-utils.ts:215-257)**:

```typescript
for (let pi = fromIdx + 1; pi <= toIdx; pi++) {
    const found = findPrefixPositionInContent(updated, pageData.content.trimStart(), searchFrom);
    // indexOf inside loop - O(n×m) worst case
}
```

For 1000-page segments, this is `1000 × 48MB = potential problem`.

**Fix**: Skip joiner normalization for segments created by the fast path (they don't need it since they use clean offset-based slicing):

```typescript
// In applyBreakpoints, after fast path:
result.push(
    ...broken.map((s) => {
        if (s.meta?.__skipJoinerNorm) { // Flag from fast path
            const { __skipJoinerNorm, ...cleanMeta } = s.meta;
            return { ...s, meta: Object.keys(cleanMeta).length > 0 ? cleanMeta : undefined };
        }
        // Apply joiner normalization (slow path only)
        // ...
    })
);
```

**Q: Concerns with synthetic data generation?**

**A**: Yes - it's too uniform. Real books have:
- Variable page lengths (1KB to 50KB)
- Dense sections with many structural markers
- Sparse sections with few markers
- Pages with only whitespace or very short content

**Improved generator**:
```typescript
function generateArabicPages(count: number): Page[] {
    const pages: Page[] = [];
    for (let i = 0; i < count; i++) {
        // Realistic variability
        const isSparse = Math.random() < 0.1; // 10% sparse pages
        const lineCount = isSparse ? 2 : 5 + Math.floor(Math.random() * 20);
        const hasMarker = !isSparse && Math.random() < 0.05; // 5% have markers
        
        const lines: string[] = [];
        if (hasMarker) {
            lines.push(`${STRUCTURAL_MARKERS[Math.floor(Math.random() * 3)]} ${generateArabicSentence(3)}`);
        }
        for (let j = 0; j < lineCount; j++) {
            const wordCount = isSparse ? 3 : 8 + Math.floor(Math.random() * 12);
            lines.push(generateArabicSentence(wordCount));
        }
        
        pages.push({ id: i + 1, content: lines.join('\n') });
    }
    return pages;
}
```

---

### 3. **Test Coverage Gaps**

**Missing critical tests**:

```typescript
describe('Fast path edge cases', () => {
    it('handles non-contiguous page IDs', () => {
        const pages: Page[] = [1, 5, 10, 100, 1000, 1001, ...].map(id => ({
            id,
            content: `Page ${id} content`,
        }));
        // Should use fast path for large span
        const result = segmentPages(pages, { maxPages: 0, breakpoints: [''], rules: [] });
        expect(result).toHaveLength(pages.length);
    });

    it('respects lineStartsAfter marker stripping in slow path', () => {
        const pages: Page[] = Array.from({ length: 100 }, (_, i) => ({
            id: i + 1,
            content: `${i + 1} - Content for page ${i + 1}`,
        }));
        const withFastPath = segmentPages(pages, {
            rules: [{ lineStartsAfter: ['{{raqms}} {{dash}} '], split: 'at' }],
            maxPages: 0,
        });
        const withoutFastPath = segmentPages(pages.slice(0, 50), { // Force slow path
            rules: [{ lineStartsAfter: ['{{raqms}} {{dash}} '], split: 'at' }],
            maxPages: 0,
        });
        // Content should be identical (markers stripped)
        expect(withFastPath[0].content).toEqual(withoutFastPath[0].content);
    });

    it('handles maxContentLength with fast path disabled', () => {
        const pages: Page[] = Array.from({ length: 2000 }, (_, i) => ({
            id: i + 1,
            content: 'A'.repeat(3000), // 3KB per page
        }));
        const result = segmentPages(pages, {
            maxContentLength: 5000,
            breakpoints: [''],
            rules: [],
        });
        expect(result.every(s => s.content.length <= 5000)).toBe(true);
    });
});
```

**Memory test skeleton**:
```typescript
it('maintains reasonable memory usage for 25k pages', () => {
    const beforeHeap = (performance as any).memory?.usedJSHeapSize ?? 0;
    const pages = generateArabicPages(25_000);
    const result = segmentPages(pages, { maxPages: 0, breakpoints: [''], rules: [] });
    const afterHeap = (performance as any).memory?.usedJSHeapSize ?? 0;
    const heapGrowthMB = (afterHeap - beforeHeap) / (1024 * 1024);
    expect(heapGrowthMB).toBeLessThan(200); // Reasonable for 25k pages
});
```

---

### 4. **Code Quality**

**Clarity Issues**:

1. **Fast path conditions are complex**: Extract to a helper:
   ```typescript
   const shouldUseFastPath = (
       pageCount: number,
       maxContentLength: number | undefined,
       debugMetaKey: string | undefined,
   ): boolean => {
       if (pageCount < FAST_PATH_THRESHOLD) return false;
       if (maxContentLength) return false; // Need character-accurate splitting
       if (debugMetaKey) return false; // Need proper provenance
       return true;
   };
   ```

2. **Magic numbers**: `10_000` and `100_000` iteration limits should be named constants with comments explaining the rationale:
   ```typescript
   /**
    * Safety limit for detecting infinite loops in breakpoint processing.
    * Normal books process in under 1000 iterations. This limit allows
    * 100x margin for edge cases while preventing hangs.
    */
   const MAX_SAFE_ITERATIONS = 100_000;
   ```

3. **Logging inside loops**: The current trace log every iteration in `processOversizedSegment` is good, but consider throttling:
   ```typescript
   if (i % 1000 === 0) { // Log every 1000 iterations
       logger?.debug?.('[breakpoints] Progress', { currentFromIdx, cursorPos, iteration: i });
   }
   ```

---

### 5. **Potential Regressions**

**High-risk scenarios**:

1. **Different output for large vs small segments**:
   ```typescript
   // This test MUST pass
   it('fast and slow paths produce identical results', () => {
       const pages = generateArabicPages(2000);
       
       // Force fast path
       const fast = segmentPages(pages, { maxPages: 0, breakpoints: [''], rules: [] });
       
       // Force slow path by processing in chunks
       const slow: Segment[] = [];
       for (let i = 0; i < pages.length; i += 100) {
           const chunk = segmentPages(
               pages.slice(i, i + 100),
               { maxPages: 0, breakpoints: [''], rules: [] }
           );
           slow.push(...chunk);
       }
       
       expect(fast).toEqual(slow); // MUST be identical
   });
   ```

2. **Replace option interaction**: If `options.replace` modifies content lengths, the fast path's offset calculations will be wrong. Current code applies replacements *before* building cumulative offsets, so this should be OK, but needs a test.

---

## Recommended Action Items

### High Priority
1. ✅ Fix page count calculation (use page ID span, not index span)
2. ✅ Add validation for cumulative offsets
3. ✅ Add test for fast/slow path equivalence
4. ✅ Add test for non-contiguous page IDs

### Medium Priority  
5. ✅ Extract fast path conditions to helper function
6. ✅ Add memory usage test
7. ✅ Improve synthetic data realism
8. ✅ Add edge case tests (empty pages, stripped markers)

### Low Priority (Nice to Have)
9. Make fast path threshold configurable
10. Add throttled logging for long-running operations
11. Document fast path limitations in AGENTS.md

---

## Final Verdict

**Is the fix correct?** Yes, with the critical page count fix.

**Is it safe?** Mostly, but needs more edge case tests.

**Is it maintainable?** Yes, with the suggested refactoring.

**Performance improvement**: Outstanding (∞ → 160ms is a massive win!)

Great work on the comprehensive documentation and test suite! The fast path is a smart optimization, and with the suggested fixes, it should be production-ready.

**Comment by ragaeeb** [comment]
# Gemini 3.0 Pro

This is a strong, pragmatic fix for a critical performance regression. By switching from  (or ) string searching to  arithmetic slicing for massive segments, you have effectively eliminated the bottleneck.

Here is my review of the changes, including a subtle edge case regarding "Offset Drift" and suggestions for your test suite.

### 1. Algorithm Correctness & Logic

The logic for `effectiveMaxPages` and the threshold check is sound.

* **The Math:** `effectiveMaxPages = maxPages + 1` correctly maps the semantic meaning of `maxPages` (span) to the number of pages per chunk.
* `maxPages: 0`  `effectiveMaxPages: 1` (1 page per chunk).
* `maxPages: 1`  `effectiveMaxPages: 2` (2 pages per chunk).


* **The Threshold:** 1000 pages is a safe heuristic. It is high enough that the overhead of string searching becomes painful, but low enough that the "Fast Path" is triggered for any substantial data dump.

### 2. The "Offset Drift" Edge Case

There is one subtle correctness trade-off in the **Fast Path** that you should be aware of.

**The Scenario:**
The Fast Path relies on `cumulativeOffsets` (derived from the original input pages) to slice `segment.content`. It assumes `segment.content` is exactly `pages.join('')`.

However, if a structural rule (like `lineStartsAfter`) ran *before* this processor and stripped markers (e.g., removed `Start:` from the first page), `segment.content` will be shorter than the sum of the pages.

**The Consequence:**
If 10 characters were stripped from the start of Page 1:

1. **Original Offsets:** Page 2 starts at index 100.
2. **Actual Content:** Page 2 starts at index 90.
3. **Fast Path:** Will still cut at index 100.
4. **Result:** Page 1's segment captures the first 10 characters of Page 2. This misalignment persists for all subsequent 24,999 segments.

**Risk Assessment:**
For the use case (recovering from a massive 25k page dump where segmentation likely failed), this is an **acceptable trade-off**. Returning *content* quickly is better than hanging while trying to find exact boundaries. However, you might want to log a warning if `segment.content.length` differs significantly from the expected cumulative length.

### 3. Missed Optimizations / Safety

You correctly identified that `maxContentLength` requires the slow path in `processOversizedSegment`. However, you **also** added a fast path to `buildBoundaryPositions` in `breakpoint-utils.ts`.

* **This is excellent.** It ensures that even if a user sets `maxContentLength` on a 25k-page book (forcing the slow loop in the processor), the *internal* boundary lookups won't hang.
* **Result:** `maxContentLength` + 25k pages will be  instead of .

### 4. Code Quality & Maintenance

* **Cleanliness:** The code is readable, and the `FAST_PATH_THRESHOLD` constant makes the intent clear.
* **Logging:** The debug logs are well-placed.
* **Files:** `breakpoint-processor.ts` and `breakpoint-utils.ts` are the correct places for this logic.

### 5. Recommendations for Test Coverage

Your performance test suite covers the "happy path" of the optimization. I recommend adding a specific correctness test for the "Offset Drift" scenario to document the behavior.

**Suggested Test Case (add to `segmenter.test.ts`):**

```typescript
it('should handle offset drift in fast-path (large segment with stripped start)', () => {
    // 1. Setup a "large" segment (mocking >1000 pages behavior with a lower threshold or mock)
    // NOTE: You might need to temporarily export/lower FAST_PATH_THRESHOLD or mock pageCount > 1000
    
    // Simpler: Just verify the behavior on a 3-page segment if you can force the fast path, 
    // but assuming you can't easily force it without 1000 pages, standard unit tests might miss this.
    
    // Instead, rely on your performance test suite to verify data integrity:
    // Ensure the SUM of lengths of output segments equals the length of the input segment.
});

```

**Specific Action for `segmenter.perf.test.ts`:**
Ensure you have a test case where `maxPages: 0` is combined with a `maxContentLength` constraint on the 25k dataset. This validates that the *secondary* fast path in `breakpoint-utils.ts` works correctly when the *primary* fast path in `breakpoint-processor.ts` is skipped.

### 6. Review Summary

| Criteria | Status | Notes |
| --- | --- | --- |
| **Correctness** | 🟢 **Pass** | Logic is sound; offset drift is acceptable for this use case. |
| **Performance** | 🟢 **Pass** |  elimination confirmed. |
| **Safety** | 🟢 **Pass** | Fallbacks for `maxContentLength` and `debugMeta` are safe. |
| **Tests** | 🟡 **Tip** | Add a check for `maxContentLength` + Large Pages to perf suite. |

### Next Step

Would you like me to write the specific **"Drift Detection"** logic (a simple heuristic to adjust `baseOffset` if `segment.content.length !== expectedLength`) to make the fast path even more accurate, or proceed with the current robust implementation?

File: src/segmentation/breakpoint-processor.ts [hunk 1]
```diff
@@ -253,11 +257,111 @@ const processOversizedSegment = (
...
+                toIdx,
+            });
+            for (let i = fromIdx; i <= toIdx; i++) {
+                const pageData = normalizedPages.get(pageIds[i]);
+                if (pageData?.content.trim()) {
+                    result.push(createSegment(pageData.content.trim(), pageIds[i], undefined, undefined));
+                }
+            }
+            return result;
+        }
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/segmentation/breakpoint-processor.ts]: https://github.com/ragaeeb/flappa-doormal/pull/16/commits/997a1350a8b8ffc2b498a00ea86815f0e4f1e724
